walreceiver: stop replication when stopping thread

kmichel-aiven opened the pull request:
This avoids failures when starting the replication twice (mainly in tests).

kmichel-aiven committed:
tests: remove unused local variable

kmichel-aiven committed:
walreceiver: stop replication when stopping thread

This avoids failures when starting the replication twice (mainly in tests).

alanfranz reviewed (dismissed):
Makes sense to me, regardless of test outcome.

alanfranz commented on pghoard/walreceiver.py:
```diff
@@ -127,6 +127,14 @@ def start_replication(self):
             self.c.start_replication(start_lsn=lsn, timeline=timeline)
         return timeline
 
+    def stop_replication(self) -> None:
```
probably there's no urgency to change this right now, but I think this method should be private (it doesn't make sense to call it from outside).

kmichel-aiven committed:
walreceiver: don't split a message in two segments

If the WAL segment buffer is almost full, it's possible for the next
message to only partially fit in the buffer.

In that case, we were sending only part of the message in the current
WAL segment, and then failing because there was something left in the
buffer.

Instead, look ahead and if we know the message won't fit, split the
WAL file before adding the message.

We will still split the message (and fail) if a single message is
larger than the WAL segment size. I don't know if that can happen
and what's the correct way to handle this case.

kmichel-aiven committed:
walreceiver: finish WAL segments on timeout

If we received no new replication message for a long time,
and some messages are still pending, flush a WAL segment.

kmichel-aiven committed:
tests: avoid concurrency bug in walreceiver

There was a possible race condition when fetching the wal_name and *then* starting walreceiver.

Fix it using a `threading.Event`.

alanfranz commented on pghoard/walreceiver.py:
```diff
@@ -117,6 +121,8 @@ def start_replication(self):
             lsn = LSN(self.last_flushed_lsn, self.pg_version_server)
         else:
             lsn = lsn_from_sysinfo(identify_system, self.pg_version_server)
+        self.initial_lsn = lsn
```
I wonder if a property-like implementation which blocks until initial_lsn is set would be better. But this looks good right now; let's think about improving it later.

alanfranz reviewed (approved):
Kudos!
